design/src/rfcs/rfc0020_service_builder.md

```diff
@@ -0,0 +1,719 @@
+- `from_handler`, which takes an async `Operation0Input -> Operation0Output`.
+
+A brief example of how this might look:
+
+```rust
+use tower::util::{ServiceFn, service_fn};
+
+impl<S> Operation0<S> {
+    pub fn from_service(inner: S) -> Self {
+        Self {
+            inner,
+        }
+    }
+}
+
+impl<F> Operation0<ServiceFn<F>> {
+    pub fn from_handler(inner: F) -> Self {
+        // Using `service_fn` here isn't strictly correct - there is slight misalignment of closure signatures. This
+        // still serves to illustrate the proposal.
+        Operation0::from_service(service_fn(inner))
+    }
+}
+```
+
+The API usage then becomes:
+
+```rust
+async fn handler(input: Operation0Input) -> Operation0Output {
+    todo!()
+}
+
+// These are both `tower::Service` and hence can have middleware applied to them
+let operation_0 = Operation0::from_handler(handler);
+let operation_1 = Operation1::from_service(/* some service */);
+
+OperationRegistryBuilder::default()
+    .operation0(operation_0)
+    .operation1(operation_1)
+    /* ... */
+```
+
+### Approach C: Operations as Middleware Constructors
+
+While [Attempt B](#approach-b-operations-as-middleware) solves all three problems, it fails to adequately model the Smithy semantics. An operation cannot uniquely define a `tower::Service` without reference to a parent Smithy service - information concerning the serialization/deserialization, error modes are all inherited from the Smithy service an operation is used within. In this way, `Operation0` should not be a standalone middleware, but become middleware once accepted by the service builder.
+
+Any solution which provides an `{Operation}` structure and wishes it to be accepted by multiple service builders must deal with this problem. We currently build one library per service and hence have duplicate structures when [service closures](https://awslabs.github.io/smithy/1.0/spec/core/model.html?highlight=closure#service-closure) overlap. This means we wouldn't run into this problem today, but it would be a future obstruction if we wanted to reduce the amount of generated code.
+
+```rust
+use tower::layer::util::{Stack, Identity};
+use tower::util::{ServiceFn, service_fn};
+
+// This takes the same form as `Operation0` defined in the previous attempt. The difference being that this is now
+// private.
+struct Service0Operation0<S> {
+    inner: S
+}
+
+impl<S> Service<http::Request> for ServiceOperation0<S>
+where
+    S: Service<Operation0Input, Response = Operation0Output, Error = Infallible>
+{
+    /* Same as above */
+}
+
+pub struct Operation0<S, L> {
+    inner: S,
+    layer: L
+}
+
+impl<S> Operation0<S, Identity> {
+    pub fn from_service(inner: S) -> Self {
+        Self {
+            inner,
+            layer: Identity
+        }
+    }
+}
+
+impl<F> Operation0<ServiceFn<F>, Identity> {
+    pub fn from_handler(inner: F) -> Self {
```

Comment:
Perhaps the omission is justified by the sentence above that said you're not going to deal with the whole handlers with and without state issue (or answered by the item in approach B "- `from_handler`, which takes an async `Operation0Input -> Operation0Output`."), but: in both this approach and approach B, what are the bounds on `F`? Is the idea to still require `F` implement `Handler`, which we will implement in codegen like we do now, for functions with and without state?

Reply:
Yep, we should continue with `Handler` pattern except the bounds change now. https://github.com/hlbarber/service-builder/pull/1/files#diff-67196527c762aa94fd3d640b4495ff46ab8e9a0760f426b9314b0e00b3f0846b